File: day15/day15.py
#!/usr/bin/env python3

import re
import sys

def parse(txt):
    pattern = re.compile(r"\w+ (-?\d+),?")
    lst = pattern.findall(txt)
    ing = []
    while lst:
        ing.append(tuple(map(int, lst[:5])))
        lst = lst[5:]
    return ing

def score(ingredients, sol):
    score = 1
    for i in range(len(ingredients[0])-1):
        p = 0
        for j, ing in enumerate(ingredients):
            p += ing[i] * sol[j]
        if p <= 0:
            return 0

        score *= p

    return score

def calories(ingredients, sol):
    return sum(ing[-1] * sol[i] for i, ing in enumerate(ingredients))

def solutions(ing):
    count = len(ing)
    sol = [0] * count
    psum = 0
    while True:
        right = count - 1
        while True:
            right -= 1
            sol[right] += 1
            psum += 1
            if psum <= 100 or right == 0:
                break
            psum -= sol[right]
            sol[right] = 0

        if sol[0] > 100:
            break
        sol[count-1] = 100 - psum
        yield sol

def solve(txt):
    part1, part2 = 0, 0
    ing = parse(txt)
    for x in solutions(ing):
        t = score(ing, x)
        if part1 < t:
            part1 = t
        if calories(ing, x) == 500 and part2 < t:
            part2 = t

    return part1, part2

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read()
    except OSError as e:
        print(e, file=sys.stderr)
        sys.exit(1)

    part1, part2 = solve(txt)
    print("Part1:", part1)
    print("Part2:", part2)

File: day15/day15.c
#include <assert.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_INGREDIENTS 16

enum
{
	CAPACITY,
	DURABILITY,
	FLAVOR,
	TEXTURE,
	CALORIES,
	MAX_PROPERTIES,
};

struct ingredient
{
	int properties[MAX_PROPERTIES];
};

struct recipe
{
	struct ingredient pool[MAX_INGREDIENTS];
	size_t count;
};

static void recipe_load(FILE *input, struct recipe *r)
{
	memset(r, 0, sizeof(*r));

	char *line = NULL;
	size_t linesize = 0;
	while (getline(&line, &linesize, input) != -1)
	{
		struct ingredient n;
		int i = 0;
		for (char *t = strtok(line, " \n,");
		     t;
		     t = strtok(NULL, " \n,"), i++)
		{
			char *end;
			if (i >= 2 && i <= 10 && (i&1) == 0)
			{
				n.properties[i / 2 - 1] = strtol(t, &end, 10);
				if (t == end || end[0] != 0)
				{
					break;
				}
			}
		}
		if (i < 10)
		{
			continue;
		}

		assert(r->count < MAX_INGREDIENTS);
		r->pool[r->count++] = n;
	}
	free(line);
}

static int recipe_score(const struct recipe *r, const int *a)
{
	int score = 1;
	for (int j = 0; j < CALORIES; j++)
	{
		int property = 0;
		for (size_t i = 0; i < r->count; i++)
		{
			property += r->pool[i].properties[j] * a[i];
		}
		if (property < 0)
		{
			property = 0;
		}

		score *= property;
	}
	return score;
}

static int recipe_calories(const struct recipe *r, const int *a)
{
	int calories = 0;
	for (size_t i = 0; i < r->count; i++)
	{
		calories += r->pool[i].properties[CALORIES] * a[i];
	}
	return calories;
}

static int next_sequence(int *a, size_t count, int *psum)
{
	size_t right = count-1;
	int sum = *psum;
	while (1)
	{
		right--;
		a[right]++;
		sum++;
		if (sum <= 100 || right == 0)
		{
			break;
		}
		sum-= a[right];
		a[right] = 0;
	}
	if (a[0] > 100)
	{
		return 0;
	}
	a[count - 1] = 100 - sum;
	*psum = sum;
	return 1;
}

static void solve(const struct recipe *r, int *ppart1, int *ppart2)
{
	int a[MAX_INGREDIENTS] = {0};
	int sum = 0;
	int part1 = INT_MIN;
	int part2 = INT_MIN;
	do
	{
		int score = recipe_score(r, a);
		if (part1 < score)
		{
			part1 = score;
		}
		if (recipe_calories(r, a) == 500
		    && part2 < score)
		{
			part2 = score;
		}
	} while (next_sequence(a, r->count, &sum));
	*ppart1 = part1;
	*ppart2 = part2;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fprintf(stderr, "Usage: %s <filename>\n", argv[0]);
		return 1;
	}

	FILE *input = fopen(argv[1], "rb");
	if (!input)
	{
		perror(argv[1]);
		return 1;
	}

	struct recipe r;
	recipe_load(input, &r);
	fclose(input);

	int part1, part2;
	solve(&r, &part1, &part2);
	printf("Part1: %d\n", part1);
	printf("Part2: %d\n", part2);
	return 0;
}
